src/odemis/cli/main.py

```diff
@@ -148,35 +148,67 @@ def print_component(comp, pretty=True, level=0):
     # * if detector, display .shape
     # * if actuator, display .axes
 
-def print_component_tree(root, pretty=True, level=0):
+
+def print_component_graph(graph, pretty=True, level=0):
     """
     Print all the components starting from the root.
-    root (Component): the component at the root of the tree
+    graph (dict Component -> dict ...): parent -> children
     pretty (bool): if True, display with pretty-printing
     level (int > 0): hierarchy level (for pretty printing)
     """
-    if pretty:
+    for comp, subg in graph.items():
         # first print the root component
-        print_component(root, pretty, level)
+        print_component(comp, pretty, level)
+        print_component_graph(subg, pretty, level + 1)
+
+
+def build_graph_children(comps):
+    """
+    comps (set)
+    return (dict comp -> dict...): parent -> children
+    """
+    # Start from the leaves, which have no children, and merge all the leaves
+    # into their parent, once the parent has all its children in the graph.
+    # Note: the children must have a single parent, otherwise, it'll not work
+    lefts = set(comps)
+    graph = {}
+    while lefts:
+        prev_lefts = lefts.copy()
+        for comp in prev_lefts:
+            children = set(comp.children.value)
+            if not (children - set(graph.keys())):
+                graph[comp] = {k: v for k, v in graph.items() if k in children}
+                for child in children:
+                    del graph[child]
+                lefts.remove(comp)
+
+        if lefts == prev_lefts:
+            logging.warning("Some components have children not in the graph: %s",
+                            ", ".join(c.name for c in lefts))
+            for comp in lefts:
+                graph[comp] = {}
+            break
+    return graph
 
-        # display all the children
-        for comp in root.children.value:
-            print_component_tree(comp, pretty, level + 1)
-    else:
-        for c in model.getComponents():
-            print_component(c, pretty)
 
 def list_components(pretty=True):
     """
     pretty (bool): if True, display with pretty-printing
     """
-    # We actually just browse as a tree the microscope
-    try:
-        microscope = model.getMicroscope()
-    except Exception:
-        raise IOError("Failed to contact the back-end")
+    # Show the root first, and don't use it for the graph, because its "children"
+    # are actually "dependencies", and it'd multiple parents in the graph.
+    microscope = model.getMicroscope()
+    print_component(microscope, pretty)
+
+    subcomps = model.getComponents() - {microscope}
```

Comment:
What is this line doing?

Reply:
Reorganised the code. Hopefully that will make it more obvious.